# 准则（必须遵守）
* 异常用于表示无法在错误发生点处理的“错误”；
* **必须**使用异常机制处理错误：如果发生无法立刻处理的“错误”，那么就使用异常机制进行报告
* 绝大多数场景中请直接使用*科力锐标准库*（**cpkt**）中“科力锐通用异常”；谨慎使用“自定义异常”，如有使用，必须在所有相关函数中进行文档注释
* 在“组件边界”必须捕获所有异常，转换为“科力锐通用异常”或“Web前后端约定的错误描述数据结构”
* 如果使用了 `except Exception`，那么就必须添加注释，注明本处代码的原由；除非是“组件边界”上的异常转换代码
* 如果`except`是为了吃掉异常，那么就必须添加注释`# do nothing，忽略掉 xxx 情况的错误`，注明这是代码编写者*有意识*的处理
* 在终结异常的时候（也就是不在继续抛出异常的时候），使用*科力锐标准库*（**cpkt**）中的方法格式化异常并打印调试日志；除非是真的不关心其错误状态，例如：使用异常进行流程控制（文档中有讨论）。

# 检查（使用异常时的坏味道）
* try是否过多：一个方法中出现了数个try；或者在一个调用层次（调用栈展开）中大多数方法都有try
* try是否过少：一个调用层次（调用栈展开）中完全没有try
* try中申请的资源是否保证离开try语块时被释放：只要没有使用with就要考虑，参考[资源释放规范](资源释放规范.md)
* 雷同的异常处理代码，需要提取为装饰器：对于同一类型的接口（如“组件边界”的调用接口），是否有雷同的异常处理；如果有就需要提取为装饰器。

# 原则（灵活使用异常机制）
* 异常机制隔离“正常逻辑”与“容错逻辑”/“意外逻辑”
* 谨慎在class之间通过异常来传递“状态”
* 谨慎使用异常机制进行流程控制，但需要考虑是否需要使用这种技巧来减少代码复杂度
* 如果有意外的异常，请完全放弃当前操作流程，不要假装知道如何处理它们
* 在“非组件边界”处检查数据有效性，且数据结构与边界处不一样时，使用assert更好
* 抑制警告的技巧
```python
try:
    ...
except Exception as e:
    _ = e # do nothing，忽略掉读取可选信息失败的错误
```

# FAQ（学习指南，以上准则与原则的由来与讨论）

## 为什么使用异常

##### 异常 与 errno
* 使用异常进行错误处理使代码更简单，更清晰，并且不太可能错过错误
* 使用 errno 导致错误处理代码与正常代码紧密相连，并且很难确保您已经处理了所有错误，参考术语： “spaghetti code”  “rat’s nest of tests”

想象一下没有异常处理会发生什么
```python
d = my_sqrt(-1) # 开方 return -1 in case of error
if d == -1:
    pass # handle error
```
```python
y = 123
x = my_negate(y) # 取反
if x == ？: # 任意整数都可能返回
    pass
```
```python
r, x = my_negate(y)
if r == -1:
    pass # handle error
```
```python
fd = os.open('some_path')
if fd == -1:
    pass # handle error
r, content = read(fd)
if r == -1:
r, dict_obj = json_loads(content)
pass # handle error
if r == -1:
    pass # handle error
```
* 要么寻找一个不可能的返回值报告错误
* 要么专门返回一个参数报告错误
* 如果某一次调用没有判断返回值，那么就会**忽略**掉错误，无论是有意的还是**无意**的

---

# 如何使用异常
* 官方文档： `https://docs.python.org/zh-cn/3.6/tutorial/errors.html`
* 注意异常捕获顺序
* 同时捕获多种异常的语法

---

# 适用/不适用使用异常的情况
##### 原则
异常机制的本意是用来进行**错误处理**（error handling）；
谨慎在非错误处理的场景下使用异常机制，大部分时候都是滥用。
##### 条例
* 用于发送“错误信号”
* 使用`except`处理你能够处理的异常
* **不要**对*代码逻辑、代码本身*引发的错误抛出异常，使用断言机制；
    * 虽然我们知道`assert`就是一个语法糖，但其含义与`raise`是不一样的。异常是“可能发生”的错误，断言是“绝对不应该发生”的错误
    * 换句话说，断言就意味着发生了需要开发人员处理的错误
```python
# 这里假定按照设计，必然能够找到index
def get_obj_index(some_obj, some_list):
    for index, list_item in enumerate(some_list):
        if is_target(some_obj, list_item):
            return index
    else:
        assert False, 'why {} NOT in {} !!!'.format(some_obj, some_list)

# 这里假定成员变量具有“不变性”
class SomeClass(object):
    def __init__(self, name):
        self.name = name
        self.some_obj = OtherClass(name)

    def some_deal(self):
        assert self.some_obj.name == self.name, 'why obj name changed !!! {} {}'.format(self.name, self.some_obj)
```
---

# 使用异常机制如何提升软件质量
###### 根源
减少了`if`分支
###### 为什么减少`if`会带来益处
尽管errno技术是常见且易懂的错误处理方式，但是添加不必要的`if`语句会带来如下问题：
* **if语句容易出错**：行业经验与数据统计得出的结论是，在其他条件类似的情况下，条件分支出错的可能性比其他类型的代码高10倍的bug率
* **if语句导致分支爆炸**：数学的指数爆炸现象可论证
后果显而易见，导致开发复杂度（复杂度）上升

---

# 为什么示例代码无法体现出异常处理的优势？
因为示例仅仅是“玩具”等级的代码，代码规模极小；而实际项目中代码规模将导致的*量变到质变*。
参考 [异常规范-附录](异常规范-附录.md)的示例
##### 优势
* 从错误的报告者，到错误的处理者；更加简洁且明确的传递错误信息
* 不需要额外的机制来将“成功”或“不成功”返回给调用者
---

# 异常将“好路径 good path”（或“快乐路径 happy path”）与“坏路径 bad path、error path”分开是什么意思？
* 快乐路径：一切顺利时的控制流路径
* 错误路径：出现问题时的控制流路径

```python
def f(): # 使用异常
    try:
        gg = g()
        hh = h()
        ii = i()
        jj = j(gg, hh, ii)
    except ErrorOneException:
        # deal error one
    except ErrorTwoException:
        # deal error two
```
可见正常情况下的逻辑路径很容易获知，处理的错误路径也明确

```python
def f(): # 使用 errno
    rc, gg = g()
    if rc == ErrorOne:
        # deal error one
        return rc
    elif rc == ErrorTwo:
        # deal error two
        return rc

    rc, hh = h()
    if rc == ErrorOne:
        # deal error one
        return rc
    elif rc == ErrorTwo:
        # deal error two
        return rc

    rc, ii = i()
    if rc == ErrorOne:
        # deal error one
        return rc
    elif rc == ErrorTwo:
        # deal error two
        return rc

    rc, jj = j(gg, hh, ii)
    if rc == ErrorOne:
        # deal error one
        return rc
    elif rc == ErrorTwo:
        # deal error two
        return rc

    return success_code
```
可以在3秒内识别出正常情况下的路径吗？

---

# 异常很简单，可以随意使用？
**No No No ！！！ 不 不 不 ！！！**
* 使用的规则更加复杂：因为返回值仅仅影响一个函数，而异常会波及整个项目
* 异常机制仅仅是工具：不合格的工程师总会把事情搞砸，无论是用errno还是异常机制
* 异常机制并非一刀切：并非所有使用errno的地方都需要用异常机制替代，有的场景下errno比异常机制更加适用
* 异常机制需要整个项目组的工程师共同维护

---

# 常见的误用异常的情况
如果使用异常机制后，导致代码更加糟糕，可能是对异常机制的滥用/误用。
*以下列举一些指导方针与思考方法，注意这些并非原则。肯定有反例的存在。*
### 使用 errno 的思维
> 其特征是对每个调用都进行 try、except。

> 但是调用后仅仅是进行业务逻辑处理，没有将异常处理掉，而是继续raise。
### 使用异常机制中的finally来保证资源回收
> 使用 with 或 context manager
### 围绕局部（子系统）定义异常
> 假设有5个子系统都可能抛出空间满异常，那么应该定义同一的异常类 `NotEnoughStorageSpace`。

> 可以想象，如果：系统A抛出 `NotEnoughSpace`，系统B抛出 `LessSpace`， 系统C抛出 `StorageFull` ...
> 那么会有额外的代码进行异常翻译，或者事实上出现更多的 if 条件结构
##### 造成这一误用的思维来自 errno 思维
> errno 为什么需要每个API独立设计？是因为可能某个接口返回3表示错误，而另外一个接口在正常的逻辑也可能返回3。
> 在实际的项目中，即便是没有异常机制的C模块，也会定义一份错误码列表，然后按照一定的规则进行返回。

> 同时由因为异常机制带来的全局化特性；在大型系统中，必须要在整个应用系统的范围考量异常的层次结构。因为异常将跨越子系统边界，成为子系统耦合的一部分。
### 使用异常中的变量来区分不同的异常
> 假设有某组方法，在3种情况下报告错误：1.文件不存在、2.文件格式不正确、3.文件被标记为无效。

> 如果外部逻辑需要分别对3种情况做出不同的错误处理。那么可以想象，如果此时仅有一种异常类型，依靠某个变量来区分到底是哪种错误。
> 那么会有额外的代码进行 if 条件筛选
### 混淆 逻辑错误 与 代码错误
> 假设有一个方法`get_user_info_by_name(name)`，发现入参为一个无效的name。
> 此时有两种可能性：1. 从用户那里收到无效的name（如用户在输入框的输入），2. 其他地方的代码错误，导致name无效
> 第一种情况适用于异常机制；第二种情况适用于断言机制。

---

# 代码中有太多的try
以下指导方针与思考方法帮助我们判读是否应该保留：
* 沿用errno思维是**不合理**
    * 特征是except异常后处理掉，通过返回值将错误返回给方法的调用者
* 吃掉异常是合理
    * 本方法中捕获异常，然后将其处理掉且不抛出任何其他异常。换句话说就是：本方法的调用者永远不知道发生过这个异常，except中不会发生错误（逻辑上）。
    * **注意：**不要吃掉不应该吃掉的异常，原则上不要吃掉Exception异常，更加不要不指定异常类型（IDE也会有警告）
    * **注意：**大部分时候应该有调试日志，记录异常已经被处理掉。详见调试日志规范
* 重新抛出捕获的异常，可能需要封装为支持with的对象，或使用装饰器
    * 特征是在捕获异常后，完成一些资源释放操作，接着重新抛出异常
    * 释放的资源就是try语块内，或紧贴着try语块
* 捕获异常，转换为新类型的异常；需要确认异常定义层次是否妥当
    * 特征：捕获XyzException，从中提取更多信息，生成PqrException对象，抛出PqrException对象

关键是为“为什么”，为什么这里在使用try

---

# 如何抛出异常
#### 不要raise Exception类型
* 因为我们不希望对 Exception 类型再进行任何判断。参考《使用异常中的变量来区分不同的异常》
#### 使用xlogging的raise_and_logging_error
* 抛出异常意味着发生了错误，大多数错误应该输出调试日志
* 默认情况下，应该抛出“通用”的异常类型

---

# 什么情况下用raise
* 记录更多调试日志
```python
def f(v):
    try:
        # ...
    except MyException as e:
        _logger.error("f failed when v = {}".format(v))
        e.add_more_debug("f failed when v = {}".format(v))
        raise
```

---

# 我应该为每种异常都定义类型么？
假设有如下需求：解析电子邮件，并提取附件内容。可能有两种风格
```python
# 使用一个常规异常，并指定错误消息：
class ParseMailError(Exception):
    pass

def get_parsed_mail(mail):
    if not mail.attachment:
        raise ParseMailError('No attachment')

    if not check_some_necessary_fields_format(mail):
        raise ParseMailError('Invalid attachment field format')

    ...
```
或者
```python
# 对于每种情况都创建一个异常类：
class ParseMailError(Exception):
    pass

class MailNoAttachmentError(ParseMailError):
    pass

class MailInvalidFieldError(ParseMailError):
    pass

def get_parsed_mail(mail):
    if not mail.attachment:
        raise MailNoAttachmentError

    if not check_some_necessary_fields_format(mail):
        raise MailInvalidFieldError

    ...
```

#### 原则
**奥卡姆剃刀**：不应该使用比必要更多的东西。

#### 场景分析
##### 如果无论什么错误，都是一样的处理
常见的仅仅是中断流程，且向用户报告错误
> 明显应该使用一个常规异常，并指定错误消息
##### 如果需要针对不同的错误，进行差异化处理
比如忽略掉无附件的情况，但是仅仅对附件内容格式有误报告
> 明显应该对于每种情况都创建一个异常类
##### 实际开发中
一般都是使用通用异常类，参考《如何抛出异常》；当发现有必要特殊处理时，添加派生类

---

# 谨慎使用异常进行流程控制
* 异常机制某些时候被类比为`超级GOTO`，谨慎使用这种技巧，可大幅简化代码。
* 如同`GOTO`一样，滥用的后果很糟糕；但合理使用，则能大幅提升代码的可读性。
* 进行流程控制必须限定在同一个方法内。
* 进行流程控制时，一定要注意捕获的异常不要超过业务逻辑的需求。
```python
try:
    orm_obj = orm_get_obj_by_value()
except ValueNotExist: # 注意不要捕获 Exception ！！！，可能导致逻辑错误
    orm_obj = orm_create_obj()
```
```python
for line in lines:
    try:
         role, line_spoken = line.split(":",1)
         print(f'role:{role} line_spoken:{line_spoken}')
    except ValueError:
         pass
```
* 禁止用异常去打断“正常”的逻辑代码。
```python
# 坏味道：用异常去打断“正常”的逻辑代码
def f(v):
    try:
        r = 'A'
        if not v:
            raise NotReallyAnException()
        r = 'A' + r
    except NotReallyAnException:
        pass
    return r
```
* 特例：简化复杂逻辑的跳出。可参考 https://blog.jooq.org/2013/04/28/rare-uses-of-a-controlflowexception/
```python
try:
    pos = find_string(haystack, needle)
    do_something_with(pos)
except NotFound: # 没找到是一种状态，但  find_string 中的逻辑非常复杂，需要从内部快速跳转到此处
    ...
```
---
